### PR TITLE
Deployment of ApiServer Secrets

### PR DESCRIPTION
**Pull request by robertgraeff**
**How to categorize this PR?**

/area delivery
/kind enhancement

**What this PR does / why we need it**:
This PR contains the deployment of the secrets for the kube api server.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
The main work is done in `kube_api_server_secrets.go` which contains the methods to create the secrets.

The example imports.yaml has been adjusted, so that `make test` creates all these secrets. (The kubeconfig needs to be added to the imports.yaml.)

**Release note**:

```other operator

```

### CODE REVIEW

**Comment by danielfoehrKn** [comment]
Looking at it today

File: pkg/virtualgarden/kube_api_server_secrets.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+	cluster := v1.NamedCluster{
+		Name: clusterName,
+		Cluster: v1.Cluster{
+			Server: "https://gardener-admission-controller.garden/webhooks/auth/seed",
+		},
+	}
+
+	if len(o.imports.VirtualGarden.KubeAPIServer.SeedAuthorizer.CertificateAuthorityData) > 0 {
+		cluster.Cluster.CertificateAuthorityData = []byte(o.imports.VirtualGarden.KubeAPIServer.SeedAuthorizer.CertificateAuthorityData)
```
Comment (danielfoehrKn) [inline pkg/virtualgarden/kube_api_server_secrets.go:160]: Does not have to be tackled in this PR. I ll open an issue.

All provided certificates should in my opinion be verified for correctness (at least that they are a valid PEM encoded x509 certificate + a CA in this case). I am doing that in the control plane component and could help implement / reference the code that's needed that at a later point.

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go:160]: Should be done in the import validation step already though.

File: pkg/virtualgarden/kube_api_server_secrets.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+			return err
+		}
+
+		// secret does not exist: generate certificate
+		cert, err := certConfig.GenerateCertificate()
+		if err != nil {
+			return err
+		}
+
+		value = cert.PrivateKeyPEM
```
Comment (danielfoehrKn) [inline pkg/virtualgarden/kube_api_server_secrets.go]: The service account RSA private key
- cannot be manually provided 
- Is not exported 

Is that on purpose? 
As a user, I would expect all the generate secrets to be exported so that I can store them safely somewhere e.g for git ops

File: pkg/virtualgarden/kube_api_server_secrets.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+			secret.Data = make(map[string][]byte)
+		}
+		secret.Data[basicAuthKey] = basicAuthValue
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	checksums[ChecksumKeyKubeAPIServerBasicAuth] = utils.ComputeChecksum(secret.Data)
```
Comment (danielfoehrKn) [inline pkg/virtualgarden/kube_api_server_secrets.go]: Only the kubeconfig containing the certificate is exported but not this one with the password. Any specific reasons for that?

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: Also, is there a reason that basic auth is enabled per default? There is already the client cert based authentication.
Mabe, we could remove it if it is not strictly required or at least make it configurable.

File: pkg/virtualgarden/kube_api_server_secrets.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *operation) deployKubeApiServerSecretAdmissionKubeconfig(ctx context.Context) error {
+	if !o.isWebhookEnabled() {
+		return nil
+	}
```
Comment (danielfoehrKn) [inline pkg/virtualgarden/kube_api_server_secrets.go:104]: is the corresponding config map  `virtual-garden-kube-apiserver-admission-config` containing the `AdmissionConfiguration` deployed at a later step?

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go:104]: Just wanting to make sure it is not forgotten.

**Review by danielfoehrKn — COMMENTED** [review]
Works fine for me. Just small remarks & questions.

**Comment by danielfoehrKn** [comment]
For me personally, the next PR can contain the whole API server deployment after this PR (so that it is testable from end-to-end).

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 1 above)
Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go:160]: Good idea. Can you send us a link to control plane coding ?

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go:160]: Will update the issue when the control plane deployment is done :)

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 2 above)
Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go]: We tried to get rid the state directory in the git repo, and store the keys and certificates only in the cluster. 
What would happen if we lose the service account RSA private key and a new one is generated?

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 4 above)
Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go:104]: Yes. Another pull request will contain the deployment of the configmaps `virtual-garden-kube-apiserver-admission-config` and `kube-apiserver-audit-policy-config`.

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 3 above)
Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go]: The basic auth credentials are currently used in the kube apiserver deployment for the [liveness and readiness probes](https://github.wdf.sap.corp/kubernetes/landscape-setup/blob/master/components/garden/virtual/charts/virtual-garden/templates/deployment-kube-apiserver.yaml#L154-L174).

Regarding the exports we followed the current deploy script. Our export parameters correspond with the data which the [current deploy script writes to the export directory](https://github.wdf.sap.corp/kubernetes/landscape-setup/blob/master/components/garden/virtual/deploy#L391-L408).

**Comment by ghost** [comment]
@robertgraeff You have mentioned internal references in the public. Please check.

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 3 above)
Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: okay thanks. Then I suppose the basic auth kubeconfig is just an "internal" config.

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 2 above)
Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: I am not sure. Would have to investigate or try out if the old service account tokens still work.

Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go]: We have installed the virtual garden, then deleted the `virtual-garden-service-account-key`secret and restarted the pods of the virtual garden. The pods (virtual-garden-kube-apiserver and virtual-garden-kube-controller-manager) did not start.

Then we executed the deployment of the virtual garden again, so that a new `virtual-garden-service-account-key`secret  was generated. The pods started, and the kube apiserver could be accessed, for example to read namespaces.

I guess that we do not need to store the secret somewhere else, since it could be recreated.

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: This private RSA key will be set in the flag `--service-account-key-file` right?
Usually, this contains the public key to verify the service account tokens generated by the KCM. In this case, it contains a private key and the API server extracts the public key from it for convenience.

If that private key is changed, I suppose the API server will fail to validate the tokens issued by the GCM.
As such this is very valuable :) 

In our current setup we supply the same RSA private key as used for the TLS serving certificate, but this is not necessary (even discouraged).

I suppose this should be exported in case someone wants to save this information.
In short: it depends whether we want to export these secrets to be stored externally. The secret itself is important.

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: I talked to @schrodit about it, and we agreed that we would be most flexible if we provide both the means to optionally import and export this RSA private key.

> We tried to get rid the state directory in the git repo, and store the keys and certificates only in the cluster.

Opened for this one: https://github.com/gardener/landscaper/issues/307
On the agenda for an upcoming sync. 
But for now, I would implement it as described above if you are fine with it.

Comment (achimweigel) [reply pkg/virtualgarden/kube_api_server_secrets.go]: I'm not sure if I understand your last paragraph correctly. Should we store the secret externally or not? As we could recreate it all the time I do not understand why?

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: > As we could recreate it all the time I do not understand why

If you re-generate the service account private key the pods running in the cluster will break on the next KAPI restart. So no, you cannot.

EDIT: I just saw that you tested it out above. 


> Should we store the secret externally or not

It is currently only stored in a secret resource (which is problematic  gardener/landscaper#307) .
I suggest for now to export it and import it.  A landscaper mechanism or a dedicated component then can take care of backing up that state and supplying it as import configuration in a subsequent run.

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: > Then we executed the deployment of the virtual garden again, so that a new virtual-garden-service-account-keysecret was generated. The pods started, and the kube apiserver could be accessed, for example to read namespaces.

That does not seem to test the relevant part. Sure the API server will fail to start if the secret is not provided & when it starts again with a new service account token key file it will be acccessible again.

Though what should be tested is whether service accounts signed by the old RSA private key in the Gardener controller manager (& verified in the API server) can still be used against the Virtual Garden API server when the service account private key has changed.
I currently do not have my own virtual garden setup to try it out.

Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go]: This means:
- We will introduce a new optional parameter for the key. If the import parameter is provided, we use it; otherwise we generate a new key.
- The key is returned in a new export parameter. (This is either the imported or the new generated key.)
- Issue [Strategy for Backup Handling #307](https://github.com/gardener/landscaper/issues/307) must clarify where the  exported value is stored and the import value comes from.

Okay?

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: I just validated that my assumption are correct on  another cluster

I updated the secret holding the service account key with another RSA private key.
Then I restart the API server.

Accessing the API with default token before rotation works

```
(shoot--dev--aws-clone:default) ~  curl -X GET https://api.aws-clone.dev.shoot.dev.k8s-hana.ondemand.com/api --header "Authorization: Bearer $TOKEN" --insecure
{
  "kind": "APIVersions",
  "versions": [
    "v1"
  ],
  "serverAddressByClientCIDRs": [
    {
      "clientCIDR": "0.0.0.0/0",
      "serverAddress": "api.aws-clone.dev.internal.dev.k8s.ondemand.com:443"
    }
  ]
}%
```

Unauthorized after the rotation

```
(shoot--dev--aws-clone:default) ~  curl -X GET https://api.aws-clone.dev.shoot.dev.k8s-hana.ondemand.com/api --header "Authorization: Bearer $TOKEN" --insecure
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}%
```

All pods mounting service account tokens are failing with

> rent data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:56.269643       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorized
2021-09-09 14:37:56.714 [INFO][1] watchercache.go 206: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:57.301366       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorized
2021-09-09 14:37:57.746 [INFO][1] watchercache.go 206: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:58.333082       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorized
2021-09-09 14:37:58.778 [INFO][1] watchercache.go 206: Failed to perform list of current data during resync ListRoot="/calico/resources/v3/projectcalico.org/nodes" error=connection is unauthorized: Unauthorized
E0909 14:37:59.364796       1 reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191114101535-6c5935290e33/tools/cache/reflector.go:96: Failed to list *v1.Node: Unauthorize


Also the API server logs are clear:

```
E0909 14:35:33.050931       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, square/go-jose: error in cryptographic primitive]"
E0909 14:35:33.052825       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, square/go-jose: error in cryptographic primitive]"
E0909 14:35:33.052828       1 authentication.go:63] "Unable to authenticate the request" err="[invalid bearer token, square/go-jose: error in cryptographic primitive]
```

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: > We will introduce a new optional parameter for the key. If the import parameter is provided, we use it; otherwise we generate a new key.
The key is returned in a new export parameter. (This is either the imported or the new generated key.)
Issue Strategy for Backup Handling #307 must clarify where the exported value is stored and the import value comes from.
Okay?


Sound good, thanks.

Comment (robertgraeff) [reply pkg/virtualgarden/kube_api_server_secrets.go]: I have implemented the changes concerning the service account key as discussed. 
I added a validation for the imported key (if it is provided).

Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: ok great, will test it later today.

File: pkg/virtualgarden/kube_api_server_secrets.go [hunk 5]
```diff
@@ -326,47 +326,47 @@ func (o *operation) generateNewEncryptionConfig() ([]byte, error) {
...
 
-	var value []byte
+	if o.imports.VirtualGarden.KubeAPIServer.ServiceAccountKeyPem == nil ||
+		len(*o.imports.VirtualGarden.KubeAPIServer.ServiceAccountKeyPem) == 0 {
 
-	secret := o.emptySecret(KubeApiServerSecretNameServiceAccountKey)
-	err := o.client.Get(ctx, client.ObjectKeyFromObject(secret), secret)
-	if err != nil {
-		if client.IgnoreNotFound(err) != nil {
-			return err
```
Comment (danielfoehrKn) [inline pkg/virtualgarden/kube_api_server_secrets.go]: Was that a miscunderstanding?
The cluster should still be checked if no import configuration is provided!
It should only generate a new one if no one was provided + not found (like before)

**Review by danielfoehrKn — CHANGES_REQUESTED** [review]

File: pkg/virtualgarden/kube_api_server_secrets.go (hunk 5 above)
Comment (danielfoehrKn) [reply pkg/virtualgarden/kube_api_server_secrets.go]: > We will introduce a new optional parameter for the key. If the import parameter is provided, we use it; otherwise we generate a new key.

Oh sorry, I have not read carefully :/

Comment (achimweigel) [reply pkg/virtualgarden/kube_api_server_secrets.go]: I have changed the coding such that a new one is only generated if no one was provided + not found

**Review by danielfoehrKn — APPROVED** [review]
